modules/apps/frontend-taglib/frontend-taglib-clay/src/main/resources/META-INF/resources/data_set_display/utils/iframes.js

```diff
@@ -13,44 +13,40 @@
  */
 
 import {OPEN_MODAL, OPEN_MODAL_FROM_IFRAME} from './eventsDefinitions';
+import {logError} from './logError';
 
-const iframeHandlerModalNamespace = 'iframe-handler-modal_';
-let modalsCounter = 0;
-const registeredModals = new Map();
+export function isPageInIframe() {
+	return window.location !== window.parent.location;
+}
 
-Liferay.on(OPEN_MODAL_FROM_IFRAME, (payload) => {
-	let firstAvailableModalId = null;
+const registeredModals = new Set();
+let modalsCounter = 0;
 
-	// eslint-disable-next-line no-for-of-loops/no-for-of-loops
-	for (const [modalId, modalRef] of registeredModals.entries()) {
-		if (modalRef) {
-			firstAvailableModalId = modalId;
-			break;
+if (!isPageInIframe()) {
+	Liferay.on(OPEN_MODAL_FROM_IFRAME, (payload) => {
+		if (!registeredModals.size) {
+			return logError('No registered modals founds');
 		}
-	}
 
-	window.top.Liferay.fire(OPEN_MODAL, {
-		...payload,
-		id: firstAvailableModalId,
+		const modalsArray = Array.from(registeredModals);
+
+		Liferay.fire(OPEN_MODAL, {
+			...payload,
+			id: modalsArray[modalsArray.length - 1],
```

Comment:
This has unnecessarily bad performance, where we are doing `O(n)` transforming a set to an array. This can be `O(1)`:
```js
const modals = registeredModals.values();

const firstAvailableModal = modals.next();
...
id: firstAvailableModal.value

```

Reply:
This won't be fixed :/ The used modal is the last one which implies a 0(n) loop anyway. 

I tried the suggested code but it wasn't working with the first one - let's wait for the dataset to be moved to another module and I'll refactor the thing by using our modal utilities ;)

Reply:
ok, sounds good :+1: